### PR TITLE
[ADD] 12.0 pos_mail_receipt module

### PR DESCRIPTION
**Pull request by PierrickBrun**

### CODE REVIEW

**Review by chienandalu — COMMENTED** [review]
Is this a backport of v13 functionality?

**Comment by PierrickBrun** [comment]
> Is this a backport of v13 functionality?

I did not know v13 included this. So this is not a backport.

This module allows you to send an e-mail to a customer without creating a partner, this is better to not force it to be more RGPD compliant.
Also in this module you send the ticket after validating the payment and not before, which is more user friendly.

**Comment by quentinDupont** [comment]
Thank you for this module !

- CSS for the email input should be changed a little to be centered

`margin-left: 50px;`

- I didn't received mail from runbot (maybe that's normal ?)

- And it was not clear if the mail was sent after clicking "OK". A green warming message "The email was sent" would be nice !

**Comment by PierrickBrun** [comment]
I "fixed" the textinput

You can see the failed sent mails in Settings -> Technical -> Email -> Emails (you need to be in debug mode)

I'll change the button color once clicked to avoid having to handle notifications (unless it is easy but I don't know how)

File: pos_mail_receipt/models/pos_order.py
Comment (legalsylvain) [inline pos_mail_receipt/models/pos_order.py:20]: I think that this function should be done asynchroneously, by queue job. 
don't you think ? 

like this module : https://github.com/OCA/pos/tree/12.0/pos_picking_delayed

File: pos_mail_receipt/static/src/js/screens.js
Comment (legalsylvain) [inline pos_mail_receipt/static/src/js/screens.js]: to remove. ;-)

File: pos_mail_receipt/models/pos_order.py
Comment (legalsylvain) [inline pos_mail_receipt/models/pos_order.py]: the returned values are not used. I think that a logger should be better. Don't you think ?

**Review by legalsylvain — COMMENTED** [review]
Some remarks inline.

FYI, here is the code used by Odoo, on master branch, to send email.

https://github.com/odoo/odoo/blob/13.0/addons/point_of_sale/models/pos_order.py#L634

General question : 
I don't understand the interest to send a receipt before the payment ? 

Thanks !

File: pos_mail_receipt/models/pos_order.py
Comment (PierrickBrun) [reply pos_mail_receipt/models/pos_order.py:20]: I'd prefer avoiding this dependency, emails are queued by default also.

I think it is better to have the e-mail sent quickly though, sometimes you need to have it to justify that you did not stole or to have a reduced parking fee.

**Comment by PierrickBrun** [comment]
@legalsylvain I've addressed the problems you saw and added tests

File: pos_mail_receipt/models/pos_order.py
Comment (legalsylvain) [reply pos_mail_receipt/models/pos_order.py:20]: > I think it is better to have the e-mail sent quickly though, sometimes you need to have it to justify that you did not stole or to have a reduced parking fee.

Hi thanks for your answer. my thought was 
- about the compatibility of this module with the certification module. (that displays the hash on the bill). But I have no time to test, and the feature was introduced in V13. (I make some developmment for V12, but not yet merged).
- IFAIK, I don't think that if you use ``send_force``, the email is not queued. I think that it will wait for the sent of the mail, and I don't think that it is in the spirit of the PoS module, to wait for server response.
ref https://github.com/OCA/pos/pull/428/files#diff-cc4623d2c77df155a2db869e209c63a2R63

But well, I don't think if it's real issues, and have no time to propose a better design, so I abstain on that point, and it's not a blocking point for me. Let's go reviewing and merging this PR, and see if it's OK.

Thanks for this work !

**Comment by PierrickBrun** [comment]
@legalsylvain @quentinDupont 
I think this is in good shape now.
I had trouble applying good CSS to the PDF generated by the backend, but in the end I got it to work

**Comment by mclaeysb** [comment]
@PierrickBrun Could you explain why the `force=True` option is used by default when the mail is sent, and when the `create_from_ui()` with `force=False` is used? I'm worried the session could be hindered when the email can't be sent.

**Comment by PierrickBrun** [comment]
> @PierrickBrun Could you explain why the `force=True` option is used by default when the mail is sent, and when the `create_from_ui()` with `force=False` is used? I'm worried the session could be hindered when the email can't be sent.

When the request is immediately sent (ie. The PoS is connected to the server), there is a `force=True` to ensure the e-mail is sent right away. This is useful because you may need it to prove you did not steal or to have a free parking for instance.
In that case, since the record is already created when we decide to send the email or not, the email is created from a separated endpoint.

When the request is sent afterwards (ie. The PoS gets back connection) there is no `force=True` because in that case it does not matter to send the email quickly.
In that case the email is sent with the same request as the record creation, in the `create_from_ui`

File: pos_mail_receipt/static/src/js/screens.js
Comment (mclaeysb) [inline pos_mail_receipt/static/src/js/screens.js]: I'm testing this branch and getting an error when the email can't be sent, stating `this.$ is not a function` referring to this line. I think this `this` should be `self`, created on line 46, as the `this` is lost here. Same goes for the `this` on line 82. This replacement fixed it for me.

**Comment by mclaeysb** [comment]
credit to @vvrossem for helping with this find!

File: pos_mail_receipt/static/src/js/screens.js
Comment (mclaeysb) [inline pos_mail_receipt/static/src/js/screens.js]: Getting the error that `_t()` is in not found. It should be declared:
```js
odoo.define(...
    var core = require('web.core');
    var _t = core._t;
```

Comment (PierrickBrun) [reply pos_mail_receipt/static/src/js/screens.js]: Hi @mclaeysb  , thanks for the review :) Would you mind committing your fixes directly to the branch ? I think you have write access to it

Comment (mclaeysb) [reply pos_mail_receipt/static/src/js/screens.js]: Sure! Sorry for the inconvenience..

**Review by flotho — APPROVED** [review]
eassy patch

LGTM

Thanks

**Comment by vdewulf** [comment]
Hello all,
Hello @mclaeysb and @PierrickBrun 
I tested the last version on one of our test database.

1) Send ticket to a customer who has an email adress.
I clicked on the send receipt button, which became green.
The email was sent correctly.

I configured some text on the header/footer in the pos config. Le sent ticket in attachment contains the modifications.

2) Selling without selecting any customer.
When I click on the button to sent the ticket, a pop up opens to indicate an email adress.
The email was sent correclty.

Something somewhat surpising : it is possible to re-click on the button, and enter a new email adress. But no ticket will be sent again.
The best would be to prevent the user to click again on the button.
But still we can live with that at the moment.

3) I cut the Wifi and made a sale.
I was able to click on the button to sent the receipt. There was no error.
The wifi signal was red with the number 1 (to inform there is 1 non synchronised sale).
I re-activated the Wifi. I clicked on the 1 number. Everything was synchronised and the email was in the queue.

=> for me everyting works fine.

**Comment by cesarlr** [comment]
We need to upload it to PyPi. Is it ok? https://github.com/coopiteasy/pos/pull/36 @mclaeysb

**Comment by mclaeysb** [comment]
Hi @cesarlr! For this PR we are waiting for the OCA (@legalsylvain @chienandalu ?) to consider it :)

File: pos_mail_receipt/static/src/css/print.css
Comment (legalsylvain) [inline pos_mail_receipt/static/src/css/print.css:53]: All this part will alter all the pdf rendered via ``web.report_assets_pdf``. Don't you think ? 
If yes it should be avoided.

File: pos_mail_receipt/models/pos_order.py
Comment (legalsylvain) [inline pos_mail_receipt/models/pos_order.py:87]: Well. Thinking regarding queue, I still think that we should do it by queue. Using queue_job for other project, it will not delay the process. Queue job are executed as soon as possible, so you'll have a delay, only if the server is busy, but the same occures with the current design.

I understand it's not in the scope of your PR, but could you add a private function ``_send_mail_receipt()`` called by ``send_mail_receipt`` with all the code. With that design, it will be easy to add a module ``pos_mail_receipt_queue_job`` inheriting

**Review by legalsylvain — CHANGES_REQUESTED** [review]
2 remarks inline. 
Could you also fixed travis. timeout is reached, just after the test you are introducing.

**Comment by legalsylvain** [comment]
some rebase should be done on this branch. (conflict, a lot of file modified, ...)

**Comment by ivantodorovich** [comment]
The branch is still messy unfortunately. Make sure you push with `--force` after rebasing.
Ideally, for this PR, only 1 commit should be here: `[ADD] pos_mail_receipt`

**Comment by flotho** [comment]
Hi @PierrickBrun , any chance to make a push rebase so that everyone could benefit from the huge amount of energy you spent on this?

Regards

**Comment by PierrickBrun** [comment]
Hi @flotho , done. I quickly tested it and it looks like nothing's been broken in the process

**Review by legalsylvain — APPROVED** [review]
Just tested again. Still think some things should be changed, but It's not a good thing to block this PR. (sorry, my bad). 

Propose to merge and let improve this module after (or wait for the odoo core feature, in more recent version).

/ocabot merge patch

**Comment by carmenbianca** [comment]
@legalsylvain This won't merge because travis is failing.

**Comment by legalsylvain** [comment]
@PierrickBrun : could you take a look on travis ? 

/ocabot rebase

**Comment by carmenbianca** [comment]
@legalsylvain I'll rebase.

**Comment by legalsylvain** [comment]
still red !

**Comment by victor-champonnois** [comment]
@legalsylvain The travis CI is red because of a timeout. Do you have an idea on  how to fix this ?

**Comment by legalsylvain** [comment]
@victor-champonnois : not sure that is a time out problem. when executing test of ``pos_mail_receipt`` something seems blocked. 

``No output has been received in the last 10m0s``...

@PierrickBrun what do you think ?

**Comment by victor-champonnois** [comment]
Note to @carmenbianca : I did not explicitly resolve the promise in the code. 
- I am not sure it is needed, since the screen is locked until the mail is sent (or raise an errors)
- Also after some trials I did not find a way to do it 
    - await does not work (I have an error telling me it must be used inside an async function/module/class, although I specifically added async to the function)
    - using .resolve() in the `email` function does not work either (`.resolve() is not a function` error)

**Comment by victor-champonnois** [comment]
> @victor-champonnois : not sure that is a time out problem. when executing test of `pos_mail_receipt` something seems blocked.
> 
> `No output has been received in the last 10m0s`...
> 
> @PierrickBrun what do you think ?

@legalsylvain You're right, apparently the problem is related to a call to `_wkhtmltopdf` in the tests.

**Comment by victor-champonnois** [comment]
@legalsylvain This is tricky. The timeout happens when the test call `send_mail_receipt()`, [itself calling `_run_wkhtmltopdf`](https://github.com/OCA/pos/blob/df17842e7cdce02eae825c1133c905dc2ae884f5/pos_mail_receipt/models/pos_order.py#L52) to convert the receipt to a pdf. 
`_run_wkhtmltopdf` run wkhtmltopdf in a subprocess. 
I tried to run the command via the command line and my conclusion is that it is stuck because of the head: 
```
                <link type="text/css" rel="stylesheet" href="/web/content/572-85cd6e9/web.report_assets_pdf.0.css"/>
                <link type="text/css" rel="stylesheet" href="/web/content/569-a6cdbb7/web.report_assets_common.0.css"/>
                <link type="text/css" rel="stylesheet" href="/web/content/570-a6cdbb7/web.report_assets_common.1.css"/>
```
When I remove it everything is fine. I think these links cannot be accessed in the test environment. 
I don't know how to fix this though. It seems impossible to remove the header this high in the call stack. 
Maybe we need to simply remove this test ?

**Comment by legalsylvain** [comment]
I dont know. @ivantodorovich any idea on that topic ?

**Comment by ivantodorovich** [comment]
I've never encountered this myself but I see something that bugs me here:
https://github.com/OCA/pos/blob/df17842e7cdce02eae825c1133c905dc2ae884f5/.travis.yml#L16

I'm not sure that's the right way of adding `wkhtmltopdf` dependency, as Odoo requires very specific versions of it.

On the oca repo template we have this: https://github.com/OCA/oca-addons-repo-template/blob/8f81ed76dd5ad0e315fb3b9729e3d7c9eeb4c867/src/%7B%25%20if%20ci%20%3D%3D%20'Travis'%20%25%7D.travis.yml%7B%25%20else%20%25%7D.t2d.yml%7B%25%20endif%20%25%7D.jinja#L67

Sadly the `12.0` branch is not using copier at all, but maybe you could try adding the travis dependency in the same way.
Not sure it'll fix the issue, but maybe it's worth trying.



> I tried to run the command via the command line and my conclusion is that it is stuck because of the head ... https://github.com/OCA/pos/pull/428#issuecomment-1164544986

Maybe `wkhtmltopdf` doesn't have access to the odoo server on the test environment? Kind of reminds me of the issues you can get due to misconfiguration and `report.url` system parameter

**Comment by victor-champonnois** [comment]
@ivantodorovich Thanks a lot, it works :)

**Comment by victor-champonnois** [comment]
(not ready to merge because new changes has been made since the approval)

File: pos_mail_receipt/static/src/js/screens.js
Comment (carmenbianca) [inline pos_mail_receipt/static/src/js/screens.js]:
```suggestion
            var timeout_db_sync = typeof options.timeout_db_sync === 'number' ? options.timeout_db_sync : 5000;
```

Alternatively, just use the same `timeout` as the line above it.

File: pos_mail_receipt/static/src/js/screens.js
Comment (carmenbianca) [inline pos_mail_receipt/static/src/js/screens.js]: This popup is not very helpful to the user.

```suggestion
                    self.gui.show_popup('error', _t('Aborted the sending of the e-mail receipt; could not synchronise with the database in time.'));
```

Comment (carmenbianca) [reply pos_mail_receipt/static/src/js/screens.js]: Also, add a `self.lock_screen(false)` line here if the `lock_screen(true)` line is moved

File: pos_mail_receipt/static/src/js/screens.js
Comment (carmenbianca) [inline pos_mail_receipt/static/src/js/screens.js]: This should be moved _before_ the loop where it checks for the DB synchronisation. The user could be waiting several seconds, and immediate feedback is good.

File: .travis.yml
Comment (ivantodorovich) [inline .travis.yml]: Actually this is how it should look like.

```suggestion
  - WKHTMLTOPDF_VERSION="0.12.5"
```

`{% - ... - %}` are jinja expressions only used by the repository template.

Adding them here unfortunately has Travis not even run because of a malformed `.travis.yml` file, so in fact we don't know yet if it's fixed or not, because CI didn't run 😢

**Review by carmenbianca — CHANGES_REQUESTED** [review]
Some comments on the recent commits by @victor-champonnois 

Awesome work :)

If you have time, make sure to run `prettier` on `screens.js`. There are a few non-idiomatic lines here and there, but it's not ultimately important.

File: pos_mail_receipt/static/src/js/screens.js
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/js/screens.js]: done, you're right, it's simpler this way.

File: pos_mail_receipt/static/src/js/screens.js
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/js/screens.js]: > This popup is not very helpful to the user.

Is "'Syncronisation with the database is taking too long. Retry or contact the system administrator.' better ? It's hard to write a helpful message since I don't know what actually *is* database syncronisation !


> Also, add a `self.lock_screen(false)` line here if the `lock_screen(true)` line is moved

done thanks :)

File: pos_mail_receipt/static/src/js/screens.js
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/js/screens.js]: To be sure I understand, do you mean adding *another* message if the state is `connecting`, before waiting for the timeout ? I am affraid that this could trigger too often...

File: pos_mail_receipt/static/src/js/screens.js
Comment (carmenbianca) [reply pos_mail_receipt/static/src/js/screens.js]: I like your proposal! Just fix the spelling to 'synchronisation' :)

(Database synchronisation, as I understand it, is the browser syncing its local cache with stuff in the DB. But I'm not 100% on the details either.)

File: pos_mail_receipt/static/src/js/screens.js
Comment (carmenbianca) [reply pos_mail_receipt/static/src/js/screens.js]: Ah, GitHub UI is doing what GitHub UI does. I'm looking exclusively at the line `self.lock_screen(true);`. That line should be moved up :)

File: pos_mail_receipt/static/src/js/screens.js
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/js/screens.js]: OK, fixed :)

File: pos_mail_receipt/static/src/js/screens.js
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/js/screens.js]: Ah I see, it's done :)

**Comment by victor-champonnois** [comment]
Status update : we didn't manage to fix the CI. We are deploying to production nevertheless.

**Comment by ivantodorovich** [comment]
> Status update : we didn't manage to fix the CI. We are deploying to production nevertheless.

Switching CI to github actions here: https://github.com/OCA/pos/pull/813
It might solve the issue

**Comment by legalsylvain** [comment]
> Switching CI to github actions here: https://github.com/OCA/pos/pull/813
> It might solve the issue

Done. Rebase (and solve conflict) required.

**Comment by ivantodorovich** [comment]
FTR, tests are still timing out even after switching to gh actions.
They should probably be re-written

**Comment by victor-champonnois** [comment]
Indeed, tests still timeout, the reason is explained [here](https://github.com/OCA/pos/pull/428#issuecomment-1164544986). But, after some time investigating, with the help of other, we still don't understand what's wrong.
@ivantodorovich 
- I don't see a way to rewrite the tests, since there is only one function to test. Do you have an idea ?
- Could we consider merging this PR without the tests ?

**Comment by ivantodorovich** [comment]
If we can't solve this issue, I would be in favour of merging without the tests.
But the situation is far from ideal, I would hate to see this issue arise in the future because some other tests trigger this method and fail. It would be a pain to debug.

I've just skimed through the code and I wonder, do we really need to call `_run_wkhtmltopdf` directly? Couldn't you just create an actual report for the `pos.order` model, and then generate+send the report when the order is saved?

**Review by remytms — APPROVED** [review]
LGTM

**Comment by remytms** [comment]
@ivantodorovich Test has been changed by @huguesdk . It's now green. :)

File: pos_mail_receipt/static/src/css/print.css
Comment (ivantodorovich) [reply pos_mail_receipt/static/src/css/print.css:53]: This comment is still pending AFAICS.

@legalsylvain has a point here, this css is not scoped so it'll be applied for any report and that's not good.

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
Thanks!! Indeed, tests are green now 🙏🏻 
Could you take a look at this comment?

File: pos_mail_receipt/models/pos_order.py
Comment (victor-champonnois) [reply pos_mail_receipt/models/pos_order.py:87]: Done

File: pos_mail_receipt/static/src/css/print.css
Comment (victor-champonnois) [reply pos_mail_receipt/static/src/css/print.css:53]: Done, I've encapsulated the CSS

**Comment by victor-champonnois** [comment]
@ivantodorovich @legalsylvain I replied to the comments, could you re-review ?

**Review by ivantodorovich — CHANGES_REQUESTED** [review]
Looks good to me 👍🏻 thanks a lot

Missing:
- CI is red, possibly a typo somewhere
- Squash your commits

**Comment by victor-champonnois** [comment]
@ivantodorovich done :)

**Review by ivantodorovich — APPROVED** [review]
Thanks!! 🙏🏻

**Comment by cesarlr** [comment]
> @ivantodorovich The merge process could not be finalized, because command `twine upload --repository-url https://upload.pypi.org/legacy/ -u __token__ odoo12_addon_pos_mail_receipt-12.0.1.0.0.99.dev7-py3-none-any.whl` failed with output:
> 
> ```
> Uploading distributions to https://upload.pypi.org/legacy/
> Uploading odoo12_addon_pos_mail_receipt-12.0.1.0.0.99.dev7-py3-none-any.whl
> �[?25l
> �[2K�[35m  0%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m0.0/36.6 kB�[0m • �[36m--:--�[0m • �[31m?�[0m
> �[2K�[35m  0%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m0.0/36.6 kB�[0m • �[36m--:--�[0m • �[31m?�[0m
> �[2K�[35m  0%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m0.0/36.6 kB�[0m • �[36m--:--�[0m • �[31m?�[0m
> �[2K�[35m  0%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m0.0/36.6 kB�[0m • �[36m--:--�[0m • �[31m?�[0m
> �[2K�[35m100%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m36.6/36.6 kB�[0m • �[33m00:00�[0m • �[31m70.9 MB/s�[0m
> �[2K�[35m100%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m36.6/36.6 kB�[0m • �[33m00:00�[0m • �[31m70.9 MB/s�[0m
> �[2K�[35m100%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m36.6/36.6 kB�[0m • �[33m00:00�[0m • �[31m70.9 MB/s�[0m
> �[2K�[35m100%�[0m �[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━�[0m �[32m36.6/36.6 kB�[0m • �[33m00:00�[0m • �[31m70.9 MB/s�[0m
> �[?25h�[33mWARNING �[0m Error during upload. Retry with the --verbose option for more details. 
> �[31mERROR   �[0m HTTPError: 403 Forbidden from https://upload.pypi.org/legacy/          
>          The user 'OCA' isn't allowed to upload to project                      
>          'odoo12-addon-pos-mail-receipt'. See                                   
>          https://pypi.org/help/#project-name for more information.              
> ```

Hi! we own this module in pypi, I will try to transfer it to OCA

**Comment by cesarlr** [comment]
I have already sent an invitation to OCA to own the project.

**Comment by ivantodorovich** [comment]
I summon @sbidoul 🙏🏻 

We have an issue here. @OCA-git-bot said the merge process failed, but [it did merge it](https://github.com/OCA/pos/tree/12.0/pos_mail_receipt) 😓 

And now it seems the module is not pushed to pypi due to this conflict.

Is the solution offered by @cesarlr enough to solve it? (inviting OCA to the pypi project)
Alternatively, what could we do here to fix this?

**Comment by sbidoul** [comment]
Is this a module that is transferred from CoopIT Easy to OCA?

**Comment by victor-champonnois** [comment]
@sbidoul Yes the PR is made from Coop IT Easy's fork

**Comment by sbidoul** [comment]
Ok, I got the PyPI invite to the OCA account. Let me see if I can accept it.

**Comment by sbidoul** [comment]
I accepted the transfer to OCA and removed coopdevs from that project on PyPI.
Please do a dummy PR to force a publish (or wait until tomorrow that the main branch bot does it).

The fact that the bot did merge anyway is tracked in https://github.com/OCA/oca-github-bot/issues/221. If anyone want's to investigate that it's welcome :)

**Pull request merged**
